src/dataframe/dataframe.jl

```diff
@@ -1508,3 +1508,63 @@ end
 function permutecols!(df::DataFrame, p::AbstractVector{Symbol})
     permutecols!(df, index(df)[p])
 end
+
+##############################################################################
+##
+## Expand & Complete dataframes
+##
+##############################################################################
+
+
+function expanddf(df::AbstractDataFrame, indexcols::Array{Symbol,1})
+    colnames = names(df)
+    # Check to make sure the symbols in indexcols are in the df
+    for i in indexcols
+        if i ∉ colnames
+            throw(ArgumentError(":$i is not in dataframe"))
+        end
+    end
+    dummydf = similar(df[indexcols],0)
+
+    # Create a dictionary of symbol=>vector of unique values
+    uniqueVals = Dict{Symbol, Array{Any,1}}()
+    for col in indexcols
+        uniqueVals[col] = unique(df[col])
+    end
+
+    # Get a long vector of every possible combination
+    collected = collect(flatten(product(values(uniqueVals)...)))
+    dfsize = div(length(collected),length(indexcols))
+
+    # Re-format to dataframe
+    for i in 1:length(indexcols):length(collected)
+        push!(dummydf, collected[i:(i+length(indexcols)-1)])
+    end
+
+    return dummydf
+end
+
+function completedf(df::AbstractDataFrame, indexcols::Array{Symbol,1}; fill=missing::Any)
+    colnames = names(df)
+    # Check to make sure the symbols in indexcols are in the df
+    for i in indexcols
+        if i ∉ colnames
+            throw(ArgumentError(":$i is not in dataframe"))
+        end
+    end
+
+    # Expand the input df and left join
+    expanded = expanddf(df, indexcols)
+    expanded = join(expanded, df; on = indexcols, kind = :left)
```

Comment:
This will produce multiple rows on duplicate combinations of values in `indexcols` - I just want to make sure we want it (or put it otherwise `expanded = join(expanded, df; on = indexcols, kind = :left)` may add rows to `expanded` data frame - do we want this?). Maybe - again - we should add a keyword argument to let the user decide on it.

Reply:
Are you referring to duplicate columns that would create duplicate combinations of values?

Reply:
I am referring to a data structure like:
```
julia> df = DataFrame(x=[1,1,1,2,2,2], y=[1,2,1,2,1,2], z=[1,2,3,4,5,6])
6×3 DataFrame
│ Row │ x     │ y     │ z     │
│     │ Int64 │ Int64 │ Int64 │
├─────┼───────┼───────┼───────┤
│ 1   │ 1     │ 1     │ 1     │
│ 2   │ 1     │ 2     │ 2     │
│ 3   │ 1     │ 1     │ 3     │
│ 4   │ 2     │ 2     │ 4     │
│ 5   │ 2     │ 1     │ 5     │
│ 6   │ 2     │ 2     │ 6     │
```

now when you do `expanddf(df, [:x, :y])` you get 4 rows, but `completedf` will produce 6 rows. And the question is if this is intended or you should throw an error in such case that duplicate were found, or you should have a keyword argument if duplicates should be allowed.

Reply:
I see what you mean. My feeling is to throw an error in the case of `completedf`, and maintain the current behavior of `expanddf`.

This also makes me consider the case provided by your example where there's already a fully expanded DF being fed in. I think `expanddf`'s behavior works as is, but as above, `completedf` should throw an error. Thoughts?

Reply:
I think it is not the issue if the table is fully epanded or not, but if it contains duplicates.
But if I understand you correctly you want `expanddf` to produce a deduplicated full factorial table and `completedf` to throw an error on duplicates. I think it is OK, but please document it.